Speculate on forceAndCall callee

JanJecmen opened the pull request.

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 2.71h
* Bounce_nonames regressed by 0.85
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1560514800&job_ids[]=1560308315&selection=all

o- commented on rir/src/compiler/opt/early_constantfold.cpp:
```diff
@@ -68,51 +70,130 @@ bool EarlyConstantfold::apply(Compiler&, ClosureVersion* cls, Code* code,
+                            ip = bb->insert(
+                                ip,
+                                new Is(
+                                    seen == SPECIALSXP
```
why not speculate on the actual callee? this would then enable inlining

o- commented on rir/src/compiler/opt/early_constantfold.cpp:
```diff
@@ -68,51 +70,130 @@ bool EarlyConstantfold::apply(Compiler&, ClosureVersion* cls, Code* code,
+                                        test, cp,
+                                        DeoptReason(
+                                            fun->typeFeedback().feedbackOrigin,
+                                            DeoptReason::Typecheck)));
```
I don't think this will work! the recordDeoptReason mechanism currently expects a typecheck to originate from a record_type_ kind of feedback. since this comes from call target recording I assume it will fail to update the typefeedback (or assert), leading to deopt loops. Or, did you actually test it?

Imho this is a callTarget deopt reason!

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_big_inline took 2.91h
* Bounce regressed by 0.93
* fasta_naive regressed by 0.93
* pidigits regressed by 0.95
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1564997649&job_ids[]=1563779276&selection=all

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_big_inline took 2.92h
* fasta_naive_2 regressed by 0.89
* reversecomplement_naive regressed by 0.93
* Overall benchmarks improved by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1566502076&job_ids[]=1563779276&selection=all

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_big_inline took 2.37h
* profiler-microbenchmark regressed by 0.93
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1572768579&job_ids[]=1563779276&selection=all

JanJecmen committed:
tests

JanJecmen committed:
refactor loading args for forceAndCall

JanJecmen committed:
add forceAndCall deopt reason

JanJecmen committed:
refactor call target feedback

JanJecmen committed:
allow stubs in mkenv

JanJecmen committed:
fix forceAndCall support

There are three modes now based on what the feedback suggests:
- monomorphic target (incl. stable body with unstable envs), we speculate on this target
- different callees but all of the same type (ie. only closures / only builtins / only specials), we speculate on the type of the target
- else we leave the forceAndCall call untouched

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_big_inline took 2.7h
* Bounce_nonames_simple regressed by 0.95
* spectralnorm_alt_3 regressed by 0.74
* nbody_naive regressed by 0.92
* matrixFor regressed by 0.9
* lapply regressed by 0.94
* Overall benchmarks regressed by 0.99

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1640144674&job_ids[]=1644533369&selection=all

JanJecmen committed:
remove unused pure flag

JanJecmen committed:
remove unused function

JanJecmen committed:
rename asLogical

JanJecmen committed:
rename llvm sexp accessors

JanJecmen committed:
typos, comments

JanJecmen committed:
rename chkClosure to chkFunction

JanJecmen committed:
split pir closure type into closure, builtin, special

JanJecmen committed:
fix typo in maybeContainsNAOrNaN

JanJecmen committed:
use common representation in Identical

JanJecmen committed:
fix unintended lambda capture
